Kenny/docs rewrite (#44)
latchbio/latch: docs/source/basics/what_is_a_workflow.md

What is a Workflow?
---

A workflow can be thought of as an analysis that takes in some input,
processes it in one or more steps and produces some output. 

> Formally, a workflow can be described as a [directed acyclic graph]() (DAG),

Reviewer: ```suggestion
Formally, a workflow is represented by a [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) (DAG),
```

where each node in the graph is called a task. This computational graph is a
flexible model to describe most any bioinformatics analysis. 

> In this example we will walk through, a workflow ingests FastQ files from a

Reviewer: ```suggestion
In this example, a workflow ingests FastQ files from a
```

sequencing machine and produces a sorted assembly. The workflow's DAG has two
tasks. The first task turns FastQ files into a BAM file using an assembly
algorithm. The second task sorts the assembly from the first task. The final
output is a useful assembly conducive to downstream analysis and visualization

> in tools like [IGV]().

Reviewer: ```suggestion
in tools like [IGV](https://software.broadinstitute.org/software/igv/).
```

documentation and a DSL to specify the presentation of parameters when the
workflow interface is generated. We'll add this content to the docstring of the
workflow function we just wrote.

```
@workflow
def assemble_and_sort(read1: LatchFile, read2: LatchFile) -> LatchFile:
    """Description...

    markdown header
    ----

    Write some documentation about your workflow in
    markdown here:

    > Regular markdown constructs work as expected.

    # Heading

    * content1
    * content2

    __metadata__:
        display_name: Assemble and Sort FastQ Files
        author:
            name:
            email:
            github:
        repository:
        license:
            id: MIT

    Args:

        read1:
          Paired-end read 1 file to be assembled.

          __metadata__:
            display_name: Read1

        read2:
          Paired-end read 2 file to be assembled.

          __metadata__:
            display_name: Read2
    """

    sam = assembly_task(read1=read1, read2=read2)
    return sort_bam_task(sam=sam)
```

## Workflow Code Structure

So far we have defined workflows and tasks as python functions but we don't know

> where to put them or what supplementary files might be needed for registration.

Reviewer 2: ```suggestion
where to put them or what supplementary files might be needed to run the code on Latch.
```

Reviewer 2: registration has not been defined

reference or boilerplate.

#### Example `Dockerfile`

**Note**: you are required to use our base image for the time being.

```
FROM 812206152185.dkr.ecr.us-west-2.amazonaws.com/wf-base:fbe8-main

# Its easy to build binaries from source that you can later reference as
# subprocesses within your workflow.
RUN curl -L https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.4.4/bowtie2-2.4.4-linux-x86_64.zip/download -o bowtie2-2.4.4.zip &&\
    unzip bowtie2-2.4.4.zip &&\
    mv bowtie2-2.4.4-linux-x86_64 bowtie2

# Or use managed library distributions through the container OS's package
# manager.
RUN apt-get update -y &&\
    apt-get install -y autoconf samtools

# You can use local data to construct your workflow image.  Here we copy a
# pre-indexed reference to a path that our workflow can reference.
COPY data /root/reference
ENV BOWTIE2_INDEXES="reference"

COPY wf /root/wf

# STOP HERE:
# The following lines are needed to ensure your build environement works
# correctly with latch.
ARG tag
ENV FLYTE_INTERNAL_IMAGE $tag
RUN  sed -i 's/latch/wf/g' flytekit.config
RUN python3 -m pip install --upgrade latch
WORKDIR /root
```

> #### Example `version` File

Reviewer: also maybe link semver
